### PR TITLE
Consistently handle `world_size==1` in the `DistributedManager`

### PR DESCRIPTION
**Pull request by akshaysubr**
# Modulus Pull Request

## Description

Closes https://github.com/NVIDIA/modulus/issues/277

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [x] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

No new dependencies

### CODE REVIEW

**Comment by akshaysubr** [comment]
@stadlmax Yes, we should go through and deal with those spots after this is merged. So far, I think AFNO, GNNs and metrics are the main ones that need to get updates.

I don't think that necessarily needs to hold up this MR through given all the tests pass. Let me know if you disagree.

File: test/distributed/test_manager.py
Comment (dallasfoster) [inline test/distributed/test_manager.py:195]: Should this be necessary if `DistributedManager.cleanup()` is used throughout?

**Review by dallasfoster — APPROVED** [review]
Looks good to me.

File: modulus/distributed/manager.py
Comment (azrael417) [inline modulus/distributed/manager.py:546]: it is good practice to specify a device id here. something like: `dist.barrier(device_ids=[<local_device_index>])`. Otherwise barriers could cause issues if more than one device is present and if there are no cuda contexts for these devices.

File: modulus/distributed/manager.py
Comment (azrael417) [inline modulus/distributed/manager.py:344]: together with line 333, does that mean that we always create a process group even if world_size = 1? If yes, how is that handled internally in torch, do all the collective ops become no-ops in that case?

File: modulus/metrics/general/ensemble_metrics.py
Comment (azrael417) [inline modulus/metrics/general/ensemble_metrics.py:193]: isn't checking against `DistributedManager().distributed` sufficient here?

File: test/metrics/test_metrics_general.py
Comment (azrael417) [inline test/metrics/test_metrics_general.py:68]: I am not sure what is best practices here, you can also compare device.type against "cuda" and device.index against 0.

**Review by azrael417 — COMMENTED** [review]
Looks good to me, I just have some minor remarks.

File: test/distributed/test_manager.py
Comment (akshaysubr) [reply test/distributed/test_manager.py:195]: No, this is not necessary. Removed.

File: modulus/distributed/manager.py
Comment (akshaysubr) [reply modulus/distributed/manager.py:546]: Agreed, added that and guarded with `if torch.cuda.is_available()`

File: modulus/distributed/manager.py
Comment (akshaysubr) [reply modulus/distributed/manager.py:344]: Looks like pytorch doesn't convert it into a no-op, but NCCL does:
```
  In [23]: with torch.profiler.profile() as p:                                                                                                                                                                                                                                                                                  
      ...:     tensor = torch.randn(20, 10, device=manager.device)                                                                                                                                                                                                                                                              
      ...:     dist.all_reduce(tensor)                                                                                                                                                                                                                                                                                          
      ...:                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                                     
  In [24]: print(p.key_averages().table(                                                                                                                                                                                                                                                                                        
      ...:     sort_by="self_cuda_time_total", row_limit=-1))                                                                                                                                                                                                                                                                   
  -------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------                                                                                                                           
                                                     Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls                                                                                                                          
  -------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------                                                                                                                           
                                            aten::normal_         0.51%      54.000us        55.02%       5.859ms       5.859ms       2.000us       100.00%       2.000us       2.000us             1                                                                                                                           
  void at::native::(anonymous namespace)::distribution...         0.00%       0.000us         0.00%       0.000us       0.000us       2.000us       100.00%       2.000us       2.000us             1                                                                                                                           
                                              aten::randn         0.22%      23.000us        97.54%      10.387ms      10.387ms       0.000us         0.00%       2.000us       2.000us             1                                                                                                                           
                                              aten::empty        42.30%       4.505ms        42.30%       4.505ms       4.505ms       0.000us         0.00%       0.000us       0.000us             1                                                                                                                           
                                    cudaStreamIsCapturing         0.10%      11.000us         0.10%      11.000us       5.500us       0.000us         0.00%       0.000us       0.000us             2                                                                                                                           
                                         cudaLaunchKernel        54.43%       5.796ms        54.43%       5.796ms       5.796ms       0.000us         0.00%       0.000us       0.000us             1                                                                                                                           
                                         c10d::allreduce_         0.54%      57.000us         1.99%     212.000us     212.000us       0.000us         0.00%       0.000us       0.000us             1                                                                                                                           
                                       record_param_comms         1.42%     151.000us         1.52%     162.000us      81.000us       0.000us         0.00%       0.000us       0.000us             2                                                                                                                           
                                      cudaStreamWaitEvent         0.08%       9.000us         0.08%       9.000us       3.000us       0.000us         0.00%       0.000us       0.000us             3                                                                                                                           
                                          nccl:all_reduce         0.00%       0.000us             0     104.000us     104.000us       0.000us         0.00%       0.000us       0.000us             1                                                                                                                           
                                    cudaDeviceSynchronize         0.40%      43.000us         0.40%      43.000us      43.000us       0.000us         0.00%       0.000us       0.000us             1                                                                                                                           
  -------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------                                                                                                                           
  Self CPU time total: 10.649ms                                                                                                                                                                                                                                                                                                 
  Self CUDA time total: 2.000us  
```

File: modulus/metrics/general/ensemble_metrics.py
Comment (akshaysubr) [reply modulus/metrics/general/ensemble_metrics.py:193]: Yes, but I'd prefer leaving this change to a separate PR to update the metrics modules to adopt these changes.

File: test/metrics/test_metrics_general.py
Comment (akshaysubr) [reply test/metrics/test_metrics_general.py:68]: I think in this particular example, comparing against the string is better because it is coming from the pytest parameterize above and so is just a string and not a `torch.device` object.

**Pull request merged**
